src/url.cpp

```diff
@@ -305,4 +331,130 @@ namespace Url
         return *this;
     }
 
+    Url& Url::relative_to(const Url& other)
+    {
+        // Support scheme-relative URLs
+        if (scheme_.empty())
+        {
+            scheme_ = other.scheme_;
+        }
+
+        // If this is an absolute URL (or scheme-relative), return early
+        if (!host_.empty()) {
+            return *this;
+        }
+
+        // If it's not an absolute URL, we need to copy the other host and port
+        host_ = other.host_;
+        port_ = other.port_;
+        userinfo_ = other.userinfo_;
+
+        // If the path portion is absolute, then bail out early.
+        if (!path_.empty() && path_.front() == '/')
+        {
+            return *this;
+        }
+
+        // Otherwise, this is a path that need to be evaluated relative to the other. If
+        // there is no '/', then we just keep our current path if it's not empty.
+        if (path_.empty())
+        {
+            if (params_.empty())
+            {
+                path_ = other.path_;
+                params_ = other.params_;
+                if (query_.empty())
+                {
+                    query_ = other.query_;
+                }
+            }
+            else
+            {
+                path_ = other.path_.substr(0, other.path_.rfind('/') + 1);
+            }
+
+            if (fragment_.empty())
+            {
+                fragment_ = other.fragment_;
+            }
+        }
+        else
+        {
+            size_t index = other.path_.rfind('/');
+            if (index != std::string::npos)
+            {
+                path_ = other.path_.substr(0, index + 1) + path_;
+            }
+            else if (!host_.empty())
+            {
+                path_ = "/" + path_;
+            }
+        }
+
+        return *this;
+    }
+
+    Url& Url::escape(bool strict)
+    {
+        escape(path_, PATH, strict);
+        escape(query_, QUERY, strict);
+        escape(params_, QUERY, strict);
+        escape(userinfo_, USERINFO, strict);
+        return *this;
+    }
+
+    void Url::escape(std::string& str, const std::string& safe, bool strict)
+    {
+        std::string copy(str);
+        size_t dest = 0;
+        // Allocate space pessimistically -- if every entity is expanded, it will take 3x
+        // the space.
+        str.resize(str.length() * 3);
+        for (size_t src = 0; src < copy.length(); ++src)
+        {
+            if (copy[src] == '%' && (copy.length() - src) >= 2)
+            {
+                // Read ahead to see if there's a valid escape sequence. If not, treat
+                // this like a normal character.
+                if (HEX_TO_DEC[copy[src+1]] != -1 && HEX_TO_DEC[copy[src+2]] != -1)
```

Comment:
Originally, I'd used `std::stoi`, but this had two drawbacks: 1) I needed to actually create another `std::string` with `copy.substr(src + 1, 2)` and 2) it throws an exception if there are no valid characters to interpret in the provided base.